dynamicscrm/metadata.go

```diff
@@ -34,109 +34,181 @@ func (c *Connector) listObjectMetadata(
 	}
 
-	return &common.ListObjectMetadataResult{
-		Result: result,
-		Errors: nil,
-	}, nil
+	fields, err := extractFields(body)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %s", err, objectName)
+	}
+
+	return fields, nil
 }
 
-// collects field properties and groups them in entities, other data in XML is ignored.
-func extractEntities(root *xquery.XML) (EntitySet, error) {
-	querySchema := fmt.Sprintf("//edmx:DataServices/Schema[@Namespace='%v']", CRMMetadataSchemaName)
-
-	schema := root.FindOne(querySchema)
-	if schema.IsEmpty() {
-		return nil, ErrMissingSchema
-	}
-
-	entities := NewEntitySet()
-	// List all field properties that exist for current schema
-	queryListAllSchemaProperties := fmt.Sprintf(
-		"//edmx:DataServices/Schema[@Namespace='%v']/EntityType[*]/Property/@Name", CRMMetadataSchemaName)
-	for _, nameAttr := range root.FindMany(queryListAllSchemaProperties) {
-		// parent of a property is an Entity.
-		// Entity may inherit properties from a parent
-		// We save entity name and the name of its parent, so later we can infer all properties by denormalisation
-		property := nameAttr.Parent()
-		entityObj := property.Parent()
-		entityName := entityObj.Attr("Name")
-		parentName := entityObj.Attr("BaseType")
-		entity := entities.GetOrCreate(entityName, parentName)
-		propertyName := property.Attr("Name")
-		entity.AddProperty(propertyName)
-	}
-
-	queryListAbstractEntities := fmt.Sprintf(
-		"//edmx:DataServices/Schema[@Namespace='%v']/EntityType[@Abstract='true']", CRMMetadataSchemaName)
-	for _, abstractEntity := range root.FindMany(queryListAbstractEntities) {
-		if !abstractEntity.HasChildren() {
-			// these entities were not included by previous query as they have no properties
-			// we programmatically find these special types, which are "primary values" but for structs
-			// Ex: crmbaseentity, crmmodelbaseentity,
-			entityName := abstractEntity.Attr("Name")
-			// effectively only create
-			_ = entities.GetOrCreate(entityName, "")
+// Given Collection of attributes from endpoint response, select those applicable to Read response.
+func extractFields(attributes *ajson.Node) (map[string]string, error) {
+	array, err := jsonquery.New(attributes).Array("value", false)
+	if err != nil {
+		return nil, errors.Join(ErrObjectNotFound, err)
+	}
+
+	if len(array) == 0 {
+		// nothing to read, we expected some attributes
+		return nil, ErrObjectMissingAttributes
+	}
+
+	fieldsMap := make(map[string]string)
+
+	for _, item := range array {
+		name, displayName, err := getAttributeNames(item)
+		if err != nil {
+			return nil, err
 		}
+
+		fieldsMap[name] = displayName
 	}
 
-	// link every child with parent completing hierarchy
-	schemaAlias := schema.Attr("Alias")
-	if err := entities.MatchParentsWithChildren(schemaAlias); err != nil {
-		return nil, errors.Join(ErrMetadataProcessing, err)
+	return fieldsMap, nil
+}
+
+func getObjectDisplayName(item *ajson.Node, objectName naming.SingularString) (string, error) {
```

Comment:
There's a lot of stuff going on here, and it might make sense to separate out these calls into another file - `entityDefinitions.go` or something - where you could summarize the steps we are going through to make metadata collection happen (the more examples, the better)

Reply:
This is a great idea. I have split methods into 2 groups. EntityDefinition vs EntityAttributes.
Metadata file uses both, with 1 main method. This should spook less people if they will open it.